src/vulkan-renderer/wrapper/swapchain.cpp

```diff
@@ -1,183 +1,289 @@
+            throw VulkanException("Error: vkQueuePresentKHR failed!", result);
+        }
     }
 }
 
-std::uint32_t Swapchain::acquire_next_image(const Semaphore &semaphore) {
-    std::uint32_t image_index = 0;
-    vkAcquireNextImageKHR(m_device.device(), m_swapchain, std::numeric_limits<std::uint64_t>::max(), semaphore.get(),
-                          VK_NULL_HANDLE, &image_index);
-    return image_index;
+bool Swapchain::is_image_usage_supported(const VkImageUsageFlagBits requested_flag,
+                                         const VkImageUsageFlags supported_flags) {
+    return (supported_flags & requested_flag) != 0u;
 }
 
-void Swapchain::recreate(std::uint32_t window_width, std::uint32_t window_height) {
-    // Store the old swapchain. This allows us to pass it to VkSwapchainCreateInfoKHR::oldSwapchain to speed up
-    // swapchain recreation.
-    VkSwapchainKHR old_swapchain = m_swapchain;
+void Swapchain::recreate(const std::uint32_t width, const std::uint32_t height) {
+    // Store the old swapchain to speed up recreation later
+    auto *const old_swapchain = m_swapchain;
+    for (auto *const img_view : m_img_views) {
+        vkDestroyImageView(m_device.device(), img_view, nullptr);
+    }
+    m_imgs.clear();
+    m_img_views.clear();
+    setup_swapchain(width, height, old_swapchain);
+}
+
+void Swapchain::setup_swapchain(const std::uint32_t width, const std::uint32_t height,
+                                const VkSwapchainKHR old_swapchain) {
+    const auto caps = m_device.get_surface_capabilities(m_surface);
+    m_surface_format = choose_surface_format(vk_tools::get_surface_formats(m_device.physical_device(), m_surface));
+    const VkExtent2D requested_extent{.width = width, .height = height};
+
+    static const std::vector<VkPresentModeKHR> default_present_mode_priorities{
+        VK_PRESENT_MODE_MAILBOX_KHR, VK_PRESENT_MODE_FIFO_RELAXED_KHR, VK_PRESENT_MODE_FIFO_KHR};
+
+    const auto composite_alpha =
+        choose_composite_alpha(VK_COMPOSITE_ALPHA_INHERIT_BIT_KHR, caps.supportedCompositeAlpha);
+
+    if (!composite_alpha) {
+        throw std::runtime_error("Error: Could not find suitable composite alpha!");
+    }
+
+    if (!is_image_usage_supported(VK_IMAGE_USAGE_COLOR_ATTACHMENT_BIT, caps.supportedUsageFlags)) {
+        throw std::runtime_error(
+            "Error: Swapchain image usage flag bit VK_IMAGE_USAGE_COLOR_ATTACHMENT_BIT is not supported!");
+    }
+
+    const auto swapchain_ci = make_info<VkSwapchainCreateInfoKHR>({
+        .surface = m_surface,
+        .minImageCount = choose_image_count(caps.minImageCount + 1, caps.minImageCount, caps.maxImageCount),
+        .imageFormat = m_surface_format.value().format,
+        .imageColorSpace = m_surface_format.value().colorSpace,
+        .imageExtent = choose_image_extent(requested_extent, caps.minImageExtent, caps.maxImageExtent, m_extent),
+        .imageArrayLayers = choose_image_array_layer_count(1, caps.maxImageArrayLayers),
```

Comment:
Don't think there's much point having the `choose_image_array_layer_count` function.

Reply:
Done.